### PR TITLE
[WIP] Rust SDK on embedded

### PR DESCRIPTION
**Pull request by gnunicorn**

### CODE REVIEW

File: _projects/rust-sdk-on-arduino.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
+---
+name: PoC: Rust SDK for arduino/embedded
+desc: Using the rust SDK to build a matrix connected embedded device
+# add a short one line description of your project
+requirements:
+# Student requirements:
+ - Knowledge on spherical astronomy.
+ - Familiar with numerical methods
+difficulty: low
```
Comment (jplatte) [inline _projects/rust-sdk-on-arduino.md:9]:
```suggestion
difficulty: hard
```

**Review by jplatte — COMMENTED** [review]
While I find the overall idea interesting, I would remove the mention of Arduino; I don't think it's realistic to get the SDK to run on bare metal at this point due to the amount of dependencies relying on the standard library as opposed to just `core` and `alloc` (unless a bare-metal target that implements enough of Rust's `std` exists, which I'm pretty sure is not the case).

**Comment by poljar** [comment]
> I added arduino mainly for the clout, but you are right, this is probably not possible at the moment. But I wonder if we need to give hands on selecting a useful embedded environment instead then or even move to less embedded-device and some iot-target instead (like a raspberry pie)... WDYT?

A Raspberry Pi can run Linux (I think it can also run Windows), I would be quite surprised if weechat-matrix-rs, or even Fractal don't run on one.

**Comment by reivilibre** [comment]
The Rust SDK already works very well on the RPi :) — I have a bot running with it.

I would be very interested to see `no_std` support in the Rust SDK so it could be used on embedded targets (ESPs might be more sensible as a first development target than Arduinos considering the memory limitations and connectivity)... Indeed I was thinking about it earlier on before I even knew about this proposal.

jplatte doesn't make it sound so trivial, though. :)

I would guess these would be involved at a minimum:

- looking at dependencies to see which ones support `no_std` environments and which ones don't. (hint: quite a few don't — `dashmap`, `lru`, ...)
- looking at how we can carve out a `no_std` subset of the crate — perhaps abstracting out things like I/O or using simplified data structures on embedded. It might be possible to contribute `no_std` support to some of the existing dependencies.
- (maybe) creating a storage layer with something embedded-friendly; e.g. LittleFS or something. I don't know how much storage the SDK actually needs, either, so there might be a difficult job there.

Things like https://lib.rs/crates/compat-no-std exist to make this kind of thing easier, but there are no silver bullets.

I would say though that this project makes *a lot of sense* considering IoT is one of the ambitions for Matrix, and this is one of the closest SDKs to bare metal we have. (though part of me wonders if it'd be easier to get nio running on MicroPython :P!)

The downsides are clear: It's unlikely to be very easy and it's not very predictable how much work it'll take.
But at least we can descope it somewhat by not including the crypto crates and other such extras for now.

**Comment by gnunicorn** [comment]
sorry, @reivilibre, seems we've posted almost at the same time.

Question for us, for now would be, do we have a mentor. Would you be interested in mentoring this project?

Sidenote: It depends on the platform, but if can get `alloc` to work, most things that are `no_std` will probably be fine still - we have it running on wasm already after all. But it probably requires some good researching on what platforms are actually feasible and still some elbow grease to get it going in a constrained environment...

**Comment by jplatte** [comment]
I *think* wasm is special in that it has a `std` that just panics at runtime for a bunch of things. I'm not certain this can be said for any bare-metal targets, but let me look for more info on this..

**Comment by jplatte** [comment]
Yeah it looks like wasm is different from bare-metal targets in that way. On the [platform support page](https://doc.rust-lang.org/nightly/rustc/platform-support.html), all three wasm targets are listed as having `std` support (even if that `std` implementation may not do everything natives ones do), while every target with `none` in its name has "std support: `*`" (only `no_std`).

There are some things that can be done to move Ruma-on-no_std (and thus matrix-sdk-on-no_std) forwards, but ultimately this is blocked on https://github.com/rust-lang/rfcs/pull/2832 and making *every* dependency that doesn't support `no_std` yet support it, or remove the need for it. See also https://github.com/ruma/ruma/issues/781.

**Pull request closed without merging**